toolchain/parse/state.def

```diff
@@ -697,45 +697,67 @@ CARBON_PARSE_STATE_VARIANTS2(ParenCondition, If, While)
 //   (state done)
 CARBON_PARSE_STATE_VARIANTS2(ParenConditionFinish, If, While)
 
-// Handles the `(` of a parenthesized expression.
+// Handles the `(` of a parenthesized single expression
 //
 // ( )
 // ^
-//   1. ParenExprFinishAsTuple
+//   1. TupleLiteralFinish
 //
 // ( ... )
 // ^
 //   1. Expr
-//   2. ParenExprParamFinishAsUnknown
-//   3. ParenExprFinishAsNormal    (SPECIAL: may be replaced)
+//   2. ExprAfterOpenParenFinish
+//   3. ParenExprFinish    (SPECIAL: may be replaced)
```

Comment:
It looks like we don't need these SPECIAL actions any more: instead, we could:

- Not push the ParenExprFinish state here
- Push either a TupleLiteralFinish state or a ParenExprFinish state in ExprAfterOpenParenFinish, depending on whether we're in the tuple case or not.

Reply:
As discussed in person yesterday, this doesn't work: we need two separate states to keep track of two separate locations (the open paren and the start of the expression inside it), unless we want to let the parser look back at already-parsed nodes for that information.

I tried revising the code to treat the two physical states as a single "logical" state that gets pushed and popped together, but that's not a straightforward change because `ConsumeListToken` expects to find the comma's parent at the top of the stack.